File: Python/basicsubscribe.py
"""Example of subscribing to an IBM MQ topic.
"""
# -*- coding: utf-8 -*-
# Copyright 2019,2025 IBM
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import logging

import ibmmq as mq

from utils.env import EnvStore

WAIT_INTERVAL = 5 # seconds

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

def connect():
    """Establish connection to the queue manager"""
    logger.info('Establishing Connection with MQ Server')
    try:
        cd = None
        if not EnvStore.ccdt_check():
            logger.info('CCDT URL export is not set, will be using json environment client connections settings')

            cd = mq.CD(Version=mq.CMQXC.MQCD_VERSION_11)
            cd.ChannelName = MQDetails[EnvStore.CHANNEL]
            cd.ConnectionName = conn_info
            cd.ChannelType = mq.CMQXC.MQCHT_CLNTCONN
            cd.TransportType = mq.CMQXC.MQXPT_TCP

            logger.info('Checking cipher details')
            # If a cipher is set then set the TLS settings
            if MQDetails[EnvStore.CIPHER]:
                logger.info('Making use of cipher details')
                cd.SSLCipherSpec = MQDetails[EnvStore.CIPHER]

        # Key repository is not specified in CCDT so look in environment settings
        # Create an empty SCO object
        sco = mq.SCO()
        if MQDetails[EnvStore.KEY_REPOSITORY]:
            logger.info('Setting Key repository')
            sco.KeyRepository = MQDetails[EnvStore.KEY_REPOSITORY]

        qmgr = mq.QueueManager(None)
        qmgr.connect_with_options(MQDetails[EnvStore.QMGR],
                                  user=credentials[EnvStore.USER],
                                  password=credentials[EnvStore.PASSWORD],
                                  cd=cd, sco=sco)
        return qmgr
    except mq.MQMIError as e:
        logger.error("Error connecting")
        logger.error(e)
        return None

def get_subscription():
    """Get access to the topic via a subscription.
    Using a MANAGED subscription means that the queue manager will create and
    return the queue to which publications are delivered.
    """
    logger.info('Creating Subscription')
    try:
        sub_desc = mq.SD()
        sub_desc.Options = mq.CMQC.MQSO_CREATE + mq.CMQC.MQSO_MANAGED
        sub_desc.set_vs("ObjectString", MQDetails[EnvStore.TOPIC_NAME])

        sub = mq.Subscription(qmgr)
        sub.sub(sub_desc=sub_desc)
        return sub
    except mq.MQMIError as e:
        logger.error("Error opening queue")
        logger.error(e)
        return None

def get_messages():
    """Get publications from the subscription queue"""
    logger.info('Attempting gets from Subscription')

    get_options = mq.CMQC.MQGMO_NO_SYNCPOINT + \
        mq.CMQC.MQGMO_FAIL_IF_QUIESCING + \
        mq.CMQC.MQGMO_WAIT + \
        mq.CMQC.MQGMO_NO_PROPERTIES

    gmo = mq.GMO(Options=get_options)
    gmo.WaitInterval = WAIT_INTERVAL * 1000 # convert to milliseconds

    # Message Descriptor
    md = mq.MD()

    keep_running = True
    while keep_running:
        try:
            # Reset the MsgId, CorrelId & GroupId so that we can reuse
            # the same 'md' object again.
            md.MsgId = mq.CMQC.MQMI_NONE
            md.CorrelId = mq.CMQC.MQCI_NONE
            md.GroupId = mq.CMQC.MQGI_NONE

            message = subscription.get(None, md, gmo)

            # Process the message here..
            logger.info('Have publication on topic: %s', message.decode())

        except mq.MQMIError as e:
            if e.comp == mq.CMQC.MQCC_FAILED and e.reason == mq.CMQC.MQRC_NO_MSG_AVAILABLE:
                # No messages, that's OK, we can just exit
                keep_running = False
            else:
                # Some other error condition.
                raise

        except (UnicodeDecodeError, ValueError) as e:
            logger.info('Message is not valid json')
            logger.info(e)
            logger.info(message)
            continue

        except KeyboardInterrupt:
            logger.info('Have received a keyboard interrupt')
            keep_running = False


def build_mq_details():
    """Create the connection details for the queue manager"""
    for key in [EnvStore.QMGR, EnvStore.TOPIC_NAME, EnvStore.CHANNEL, EnvStore.HOST,
                EnvStore.PORT, EnvStore.KEY_REPOSITORY, EnvStore.CIPHER]:
        MQDetails[key] = EnvStore.getenv_value(key)


# Application Logic starts here
logger.info("Application 'BasicSubscribe' is starting")

envStore = EnvStore()
envStore.set_env()

MQDetails = {}
credentials = {
    EnvStore.USER: EnvStore.getenv_value(EnvStore.APP_USER),
    EnvStore.PASSWORD: EnvStore.getenv_value(EnvStore.APP_PASSWORD)
}

build_mq_details()

logger.info('Credentials are set')
conn_info = EnvStore.get_connection(EnvStore.HOST, EnvStore.PORT)

qmgr = None
subscription = None

qmgr = connect()
if qmgr is not None:
    subscription = get_subscription()
if subscription is not None:
    get_messages()
    subscription.close(close_sub_queue=True)

if qmgr is not None:
    qmgr.disconnect()

logger.info("Application is ending")

File: Node.js/basicsubscribe.js
/**
 * Copyright 2019 IBM Corp.
 *
 * Licensed under the Apache License, Version 2.0 (the 'License');
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 **/


// This is a demonstration showing the basic subscribe operations onto a MQ Topic
// Using the MQI Node.js interface

// This application is based on the samples
// https://github.com/ibm-messaging/mq-mqi-nodejs/blob/master/samples/amqsconn.js
// and
// https://github.com/ibm-messaging/mq-mqi-nodejs/blob/master/samples/amqssub.js
//
// Values for Queue Manager, Topic, Host, Port and Channel are
// passed in as envrionment variables.

const fs = require('fs');
// Import the MQ package
var mq = require('ibmmq');
// Import any other packages needed
var StringDecoder = require('string_decoder').StringDecoder;
var decoder = new StringDecoder('utf8');

// Load up missing envrionment variables from the env.json file
var env = require('../env.json');

var MQC = mq.MQC; // Want to refer to this export directly for simplicity

// Set up debug logging options
var debug_info = require('debug')('amqssub:info');
var debug_warn = require('debug')('amqssub:warn');

// Set up Constants
const CCDT = "MQCCDTURL";
const	FILEPREFIX = "file://";

// Load the MQ Endpoint details either from the envrionment or from the
// env.json file. The envrionment takes precedence. The json file allows for
// mulitple endpoints ala a cluster.
// A connection string is built using each endpoint's HOST(PORT)
var MQDetails = {};

['QMGR', 'TOPIC_NAME', 'HOST', 'PORT',
 'CHANNEL', 'KEY_REPOSITORY', 'CIPHER'].forEach(function(f) {
  MQDetails[f] = process.env[f] || env.MQ_ENDPOINTS[0][f]
});

var credentials = {
  USER: process.env.APP_USER || env.MQ_ENDPOINTS[0].APP_USER,
  PASSWORD: process.env.APP_PASSWORD || env.MQ_ENDPOINTS[0].APP_PASSWORD
}

// Global variables
var ok = true;


function getConnection() {
  let points = [];
  env.MQ_ENDPOINTS.forEach((p) => {
    if (p['HOST'] && p['PORT']) {
      points.push(`${p.HOST}(${p.PORT})`)
    }
  });

  return points.join(',');
}

// Define some functions that will be used from the main flow
function getMessages(hObj) {
  while (ok) {
    getMessage(hObj);
  }
}

// This function retrieves messages from the queue without waiting using
// the synchronous method for simplicity. See amqsgeta for how to use the
// async method.
function getMessage(hObj) {

  var buf = Buffer.alloc(1024);

  var mqmd = new mq.MQMD();
  var gmo = new mq.MQGMO();

  //gmo.WaitInterval = 3 * 1000; // 3 seconds
  gmo.WaitInterval = 30 * 1000; // 30 seconds
  gmo.Options = MQC.MQGMO_NO_SYNCPOINT |
    MQC.MQGMO_WAIT |
    MQC.MQGMO_CONVERT |
    MQC.MQGMO_FAIL_IF_QUIESCING;

  mq.GetSync(hObj, mqmd, gmo, buf, function(err, len) {
    if (err) {
      if (err.mqrc == MQC.MQRC_NO_MSG_AVAILABLE) {
        debug_info("no more messages");
      } else {
        debug_warn("MQGET failed with " + err.mqrc);
      }
      ok = false;
    } else if (mqmd.Format == "MQSTR") {
      var msgObject = null;
      try {
        msgObject = JSON.parse(buf);
        debug_info("Message Object found", msgObject);
      } catch (err) {
        debug_info("message <%s>", decoder.write(buf));
      }
    } else {
      debug_info("binary message: " + buf);
    }
  });
}

// When we're done, close queues and connections
function cleanup(hConn, hObjPubQ, hObjSubscription) {
  // Demonstrate two ways of closing queues - first using an exception, then
  // the version with callback.
  try {
    mq.Close(hObjSubscription, 0);
    debug_info("MQCLOSE (Subscription) successful");
  } catch (err) {
    debug_warn("MQCLOSE (Subscription) ended with reason " + err.mqrc);
  }

  mq.Close(hObjPubQ, 0, function(err) {
    if (err) {
      debug_warn("MQCLOSE (PubQ) ended with reason " + err.mqrc);
    } else {
      debug_info("MQCLOSE (PubQ) successful");
    }
    mq.Disc(hConn, function(err) {
      if (err) {
        debug_warn("MQDISC ended with reason " + err.mqrc);
      } else {
        debug_info("MQDISC successful");
      }
    });
  });
}

function ccdtCheck () {
  if (CCDT in process.env) {
    let ccdtFile = process.env[CCDT].replace(FILEPREFIX, '');
    debug_info(ccdtFile);
    if (fs.existsSync(ccdtFile)) {
      debug_info("CCDT File found at ", ccdtFile);
      return true;
    }
  }
  return false;
}

// The program really starts here.
// Connect to the queue manager. If that works, the callback function
// opens the topic, and then we can start to retrieve messages.

debug_info('Running on ', process.platform);
debug_info('Starting up Application');

var cno = new mq.MQCNO();
// use MQCNO_CLIENT_BINDING to connect as client
// cno.Options = MQC.MQCNO_NONE;
cno.Options = MQC.MQCNO_CLIENT_BINDING;

// For no authentication, disable this block
if (credentials.USER) {
  var csp = new mq.MQCSP();
  csp.UserId = credentials.USER;
  csp.Password = credentials.PASSWORD;
  cno.SecurityParms = csp;
}

if (! ccdtCheck()) {
  debug_info('CCDT URL export is not set, will be using json envrionment client connections settings');

  // And then fill in relevant fields for the MQCD
  var cd = new mq.MQCD();

  cd.ChannelName = MQDetails.CHANNEL;
  cd.ConnectionName = getConnection();
  debug_info('Connections string is ', cd.ConnectionName);

  if (MQDetails.KEY_REPOSITORY) {
    debug_info('Will be running in TLS Mode');

    cd.SSLCipherSpec = MQDetails.CIPHER;
    cd.SSLClientAuth = MQC.MQSCA_OPTIONAL;
  }

  // Make the MQCNO refer to the MQCD
  cno.ClientConn = cd;
}

// The location of the KeyRepository is not specified in the CCDT, so regardless
// of whether a CCDT is being used, need to specify the KeyRepository location
// if it has been provided in the environment json settings.
if (MQDetails.KEY_REPOSITORY) {
  debug_info('Key Repository has been specified');
  // *** For TLS ***
  var sco = new mq.MQSCO();

  sco.KeyRepository = MQDetails.KEY_REPOSITORY;
  // And make the CNO refer to the SSL Connection Options
  cno.SSLConfig = sco;
}


// Do the connect, including a callback function
mq.Connx(MQDetails.QMGR, cno, function(err, hConn) {
  if (err) {
    console.log("MQCONN ended with reason code " + err.mqrc);
  } else {
    console.log("MQCONN to %s successful ", MQDetails.QMGR);

    // Define what we want to open, and how we want to open it.
    var sd = new mq.MQSD();
    sd.ObjectString = MQDetails.TOPIC_NAME;
    sd.Options = MQC.MQSO_CREATE |
      MQC.MQSO_NON_DURABLE |
      MQC.MQSO_FAIL_IF_QUIESCING |
      MQC.MQSO_MANAGED;

    mq.Sub(hConn, null, sd, function(err, hObjPubQ, hObjSubscription) {
      if (err) {
        debug_warn("MQSUB ended with reason " + err.mqrc);
      } else {
        debug_info("MQSUB to topic %s successfull", MQDetails.TOPIC_NAME);
        // And loop getting messages until done.
        getMessages(hObjPubQ);
      }
      cleanup(hConn, hObjPubQ, hObjSubscription);
    });
  }
});
